### PR TITLE
5326 close prompts on load

### PR DESCRIPTION
**Pull request by Baron-Severin**
---

### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- [ ] **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- [ ] **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

File: components/feature/prompts/src/test/java/mozilla/components/feature/prompts/PromptFeatureTest.kt [hunk 1]
```diff
@@ -844,3 +856,153 @@ class PromptFeatureTest {
...
+                url = "www.wikipedia.org",
+                loading = false,
+                progress = 100
+            ))
+        )
+
+        dismissOnPageMostlyLoaded(dialog, stateFlow)
+
+        verify(dialog, times(1)).dismiss()
+    }
```
Comment (Baron-Severin) [inline components/feature/prompts/src/test/java/mozilla/components/feature/prompts/PromptFeatureTest.kt]: TODO add test ensures the fun does nothing if the scope has already been canceled.

File: components/feature/prompts/src/test/java/mozilla/components/feature/prompts/PromptFeatureTest.kt [hunk 2]
```diff
@@ -836,6 +834,70 @@ class PromptFeatureTest {
...
+        store.dispatch(ContentAction.UpdateProgressAction(tabId, 11)).joinBlocking()
+        store.dispatch(ContentAction.UpdateProgressAction(tabId, 28)).joinBlocking()
+        store.dispatch(ContentAction.UpdateProgressAction(tabId, 32)).joinBlocking()
+        store.dispatch(ContentAction.UpdateProgressAction(tabId, 49)).joinBlocking()
+        store.dispatch(ContentAction.UpdateProgressAction(tabId, 60)).joinBlocking()
+        store.dispatch(ContentAction.UpdateProgressAction(tabId, 89)).joinBlocking()
+        store.dispatch(ContentAction.UpdateProgressAction(tabId, 90)).joinBlocking()
+
+        // dismiss + clear
+        verify(feature, times(2)).activePrompt
```
Comment (Baron-Severin) [inline components/feature/prompts/src/test/java/mozilla/components/feature/prompts/PromptFeatureTest.kt]: I couldn't figure out any better ways to sense `dismiss` calls in these tests.  I'm open to other options, if anyone has any ideas.

File: components/feature/prompts/src/main/java/mozilla/components/feature/prompts/PromptFeature.kt [hunk 3]
```diff
@@ -108,11 +112,15 @@ class PromptFeature private constructor(
     onNeedToRequestPermissions: OnNeedToRequestPermissions
 ) : LifecycleAwareFeature, PermissionsFeature, Prompter {
     private var scope: CoroutineScope? = null
+    private var scope2: CoroutineScope? = null
```
Comment (csadilek) [inline components/feature/prompts/src/main/java/mozilla/components/feature/prompts/PromptFeature.kt]: @Baron-Severin let's give these scopes better names :)

File: components/feature/prompts/src/test/java/mozilla/components/feature/prompts/PromptFeatureTest.kt [hunk 4]
```diff
@@ -836,6 +834,70 @@ class PromptFeatureTest {
...
+    }
+
+    @Test
+    fun `dialog will be dismissed if progress skips past 90%`() {
```
Comment (csadilek) [inline components/feature/prompts/src/test/java/mozilla/components/feature/prompts/PromptFeatureTest.kt]: nit: maybe `if new page load progress`.

File: components/feature/prompts/src/test/java/mozilla/components/feature/prompts/PromptFeatureTest.kt (hunk 2 above)
Comment (csadilek) [inline components/feature/prompts/src/test/java/mozilla/components/feature/prompts/PromptFeatureTest.kt]: We can make `activePrompt` internal so we can access from the test and set and mock. This is a pattern we use commonly, see `promptAbuseDetector`.

File: components/feature/prompts/src/main/java/mozilla/components/feature/prompts/PromptFeature.kt (hunk 3 above)
Comment (Baron-Severin) [reply components/feature/prompts/src/main/java/mozilla/components/feature/prompts/PromptFeature.kt]: Whoops!  I meant to comment there too...

File: components/feature/prompts/src/main/java/mozilla/components/feature/prompts/PromptFeature.kt [hunk 5]
```diff
@@ -205,6 +214,17 @@ class PromptFeature private constructor(
                 }
         }
 
+        @Suppress("MagicNumber")
```
Comment (csadilek) [inline components/feature/prompts/src/main/java/mozilla/components/feature/prompts/PromptFeature.kt]: nit: Let's make `90` a constant?

File: components/feature/prompts/src/main/java/mozilla/components/feature/prompts/PromptFeature.kt [hunk 6]
```diff
@@ -205,6 +214,17 @@ class PromptFeature private constructor(
...
+        dismissPromptScope = store.flowScoped { flow ->
```
Comment (csadilek) [inline components/feature/prompts/src/main/java/mozilla/components/feature/prompts/PromptFeature.kt]: I'd also add a comment here why were doing this as it's not obvious. Basically to dismiss popups if a new page page loads while the popup displays, or maybe link to the issue...

**Comment by Baron-Severin** [comment]
bors r=csadilek

**Pull request merged**